src/subscription_manager/syspurposelib.py

```diff
@@ -216,5 +237,105 @@ def read_syspurpose():
+            if raise_on_error:
+                raise
             syspurpose = {}
     return syspurpose
+
+
+def write_syspurpose(values):
+    """
+    Write the syspurpose to the file system.
+    :param values:
+    :return:
+    """
+    if SyspurposeStore is not None:
+        sp = SyspurposeStore(USER_SYSPURPOSE)
+        sp.contents = values
+        sp.write()
+    else:
+        # Simple backup in case the syspurpose tooling is not installed.
+        try:
+            json.dump(values, open(USER_SYSPURPOSE), ensure_ascii=True, indent=2)
+        except OSError:
+            log.warning('Could not write syspurpose to %s' % USER_SYSPURPOSE)
+            return False
+    return True
+
+
+class SyspurposeSyncActionInvoker(certlib.BaseActionInvoker):
+    """
+    Used by rhsmcertd to sync the syspurpose values locally with those from the Server.
+    """
+
+    def _do_update(self):
+        action = SyspurposeSyncActionCommand()
+        return action.perform()
+
+
+class SyspurposeSyncActionReport(certlib.ActionReport):
+    name = "Syspurpose Sync"
+
+
+class SyspurposeSyncActionCommand(object):
+    """
+    Sync the system purpose values, by performing a three-way merge between:
+      - The last known shared state (SyspurposeCache)
+      - The current values on the server
+      - The current values on the file system
+    """
+
+    def __init__(self):
+        self.report = SyspurposeSyncActionReport()
+        self.cp_provider = inj.require(inj.CP_PROVIDER)
+        self.uep = self.cp_provider.get_consumer_auth_cp()
+
+    def perform(self):
+        """
+        Perform the action that this Command represents.
+        :return:
+        """
+        self.sync()
+        return self.report
+
+    def sync(self):
+        """
+        Actually do the sync between client and server.
+        Saves the merged changes between client and server in the SyspurposeCache.
+        :return: The synced values
+        """
+        if not self.uep.has_capability('syspurpose'):
+            log.debug('Server does not support syspurpose, not syncing')
+            return
+
+        consumer_identity = inj.require(inj.IDENTITY)
+        consumer = self.uep.getConsumer(consumer_identity.uuid)
+
+        server_sp = {}
+        sp_cache = SyspurposeCache()
+        # Translate from the remote values to the local, filtering out items not known
+        for attr in ATTRIBUTES:
+            server_sp[attr] = consumer.get(LOCAL_TO_REMOTE[attr])
+
+        try:
+            filesystem_sp = read_syspurpose(raise_on_error=True)
+        except (os.error, ValueError):
+            self.report._exceptions.append(
+                    'Cannot read local syspurpose, trying to update from server only'
+            )
+            result = server_sp
+        else:
+            cached_values = sp_cache.read_cache_only()
+            result = three_way_merge(local=filesystem_sp, base=cached_values, remote=server_sp)
```

Comment:
Just proposal: It would be nice to see in `log.debug()` how was `result` created (from what sources: local/base/remote). I can imagine customers confused that content of `/etc/rhsm/syspurpose/syspurpose.json` is periodically and 'magically' changed.